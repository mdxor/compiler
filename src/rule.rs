use regex::Regex;

// pub struct Rule {
//   pub thematic_break: Regex,
//   pub atx_heading: Regex,
//   pub empty_atx_heading: Regex,
//   pub indented_code: Regex,
//   pub closing_atx_heading: Regex,
//   pub fenced_code: Regex,
//   pub fenced_code_end: Regex,
//   pub whitespace: Regex,
//   pub blank_line: Regex,
//   pub line: Regex,
// }

// impl Rule {
//   pub fn new() -> Self {
//     let line_ending = "(?:\n|$)";
//     Rule {
//       thematic_break: Regex::new(
//         "^(?:(?:\\* *){3,}|(?:_ *){3,}|(?:\\- *){3,})line_ending"
//           .replace("line_ending", line_ending),
//       )
//       .unwrap(),
//       indented_code: Regex::new("^ {4}").unwrap(),
//       atx_heading: Regex::new("^(#{1,6}) ").unwrap(),
//       empty_atx_heading: Regex::new("^(#{1,6})(?:\n|$)").unwrap(),
//       closing_atx_heading: Regex::new("(^| )#+ *$").unwrap(),
//       fenced_code: Regex::new("^```").unwrap(),
//       fenced_code_end: Regex::new("(^ {0,3}|\n {0,3})``` *\n?").unwrap(),
//       whitespace: Regex::new("^ +").unwrap(),
//       blank_line: Regex::new("^ +\n").unwrap(),
//       line: Regex::new("^(.*)(?:\n|$)").unwrap(),
//     }
//   }
// }
