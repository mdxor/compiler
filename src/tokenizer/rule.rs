pub const ATX_HEADING_RULE: &str = "^(#{1,6})(?:| (.*))(?:\n|$)";
pub const ATX_HEADING_CLOSING_RULE: &str = "(^| )#+ *$";
pub const THEMATIC_BREAK_RULE: &str = r"((?:- *){3,}|(?:_ *){3,}|(?:\* *){3,})(?:\n+|$)";
