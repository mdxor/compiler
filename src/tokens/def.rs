#[cfg(test)]
use serde::Serialize;

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct ATXHeading<'source> {
  pub level: usize,
  pub raw_inlines: &'source str,
}
