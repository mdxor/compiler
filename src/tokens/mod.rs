mod atx_heading;
mod def;
mod rule;
