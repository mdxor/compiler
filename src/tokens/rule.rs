pub const ATX_HEADING_RULE: &str = "^(#{1,6})(?:| (.*))(?:\n|$)";
