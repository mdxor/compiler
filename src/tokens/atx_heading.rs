use crate::tokens::def::*;
use crate::tokens::rule;
use lazy_static::lazy_static;
use regex::Regex;
pub fn atx_heading(source: &str) -> Option<ATXHeading> {
  lazy_static! {
    static ref RULE: Regex = Regex::new(rule::ATX_HEADING_RULE).unwrap();
  }
  if let Some(caps) = RULE.captures(source) {
    let level = caps.get(1).unwrap().as_str().len();
    let raw_inlines = caps.get(2).map_or("", |v| v.as_str());
    Some(ATXHeading { level, raw_inlines })
  } else {
    None
  }
}

#[test]
fn test_atx_heading() {
  let cases = vec![
    "# foo",
    "## foo",
    "### foo",
    "#### foo",
    "##### foo",
    "###### foo",
    "####### foo",
    "#5 bolt",
    "#hashtag",
    r"\## foo",
    r"# foo *bar* \*baz\*",
    "###",
  ];
  let mut results = vec![];
  for case in cases {
    results.push(atx_heading(case));
  }
  insta::assert_yaml_snapshot!(results);
}
