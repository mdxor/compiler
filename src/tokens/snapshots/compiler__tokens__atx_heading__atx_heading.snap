---
source: src/tokens/atx_heading.rs
expression: results

---
- level: 1
  raw_inlines: foo
- level: 2
  raw_inlines: foo
- level: 3
  raw_inlines: foo
- level: 4
  raw_inlines: foo
- level: 5
  raw_inlines: foo
- level: 6
  raw_inlines: foo
- ~
- ~
- ~
- ~
- level: 1
  raw_inlines: "foo *bar* \\*baz\\*"
- level: 3
  raw_inlines: ""

